feat: add open in coder docs, fix missing templates

jsjoeio opened the pull request:
## Changes

- adds docs for the first version of the Open in Coder button 
- fix template missing error
- add Open in Coder button to dogfood `README`

### Screenshots

This is what the new error looks like on the Template page
![erro-template](https://user-images.githubusercontent.com/3806031/191358531-67a88fe8-b2d2-4bc5-8daf-79dac80d3d55.jpg)


Part 1 of ? for https://github.com/coder/coder/issues/3981

## Changes to address

- [x] remove query param on index page
- [x] change docs
- [x] add example to the dogfood docs
- [x] update PR description

jsjoeio committed:
docs: add open in coder

This adds new documentation for the "Open in Coder" button that admins
can use to get their developers up and running faster.

jsjoeio commented:
@bpmct would love early feedback on these docs. Also, do we have an "Open in Coder" image asset we can use?

bpmct commented:
> would love early feedback on these docs. Also, do we have an "Open in Coder" image asset we can use?

Yep, we have https://cdn.coder.com/embed-button.svg

bpmct commented on docs/admin/open-in-coder.md:
```diff
@@ -0,0 +1,29 @@
+# Open in Coder Button
+
+This is a Markdown button you can add to your project's `README.md` to get your developers up and running with Coder twice as fast.
```
```suggestion
Add a markdown button to your project's `README.md` to get your developers up and running with Coder with a few clicks.
```

bpmct commented on docs/admin/open-in-coder.md:
```diff
@@ -0,0 +1,29 @@
+
+These are the following query parameters we support:
+
+### template
```
```suggestion
### Template
```

bpmct commented on docs/admin/open-in-coder.md:
````diff
@@ -0,0 +1,29 @@
+A basic example looks like this:
+
+```text
+[![open in coder button](https://cdn.coder.com/embed-button.svg)](https://dev.coder.com/?template=coder-ts)
````
```suggestion
[![Open in Coder](https://cdn.coder.com/embed-button.svg)](https://dev.coder.com/?template=coder-ts)
```

bpmct commented on docs/admin/open-in-coder.md:
```diff
@@ -0,0 +1,29 @@
+
+which renders like this:
+
+![open in coder button](https://cdn.coder.com/embed-button.svg)
```
```suggestion
![Open in Coder](https://cdn.coder.com/embed-button.svg)
```

bpmct commented on docs/admin/open-in-coder.md:
```diff
@@ -0,0 +1,29 @@
+
+The underlying link for this button consists of the following pieces:
+- <deployment-url>: where your Coder deployment lives i.e. https://dev.coder.com
+- <query-params>: optional query parameters to customize the experience
```
Why not add each query parameter here?

jsjoeio replied:
As in replace `<query-params>` with each query param here? I think I know what you're saying but feel free to add a suggestion!

jsjoeio committed:
fix: display error if template not found

Previously, we weren't handling a case where we tried to get a template
that returned a 404 from the backend.

Now we handle that case in our state machine and display the error
message from the API on the frontend.

jsjoeio committed:
feat: support template query param in index

This adds support to navigate directly to a template from the index by
using the `?template=<name>` query  param.

jsjoeio commented:
@bpmct This completes the work for support of the `template` query param. My thinking is to merge this and then I add all the other query params in subsequent PRs.

kylecarbs commented on docs/admin/open-in-coder.md:
````diff
@@ -0,0 +1,29 @@
+A basic example looks like this:
+
+```text
+[![Open in Coder](https://cdn.coder.com/embed-button.svg)](https://dev.coder.com/?template=coder-ts)
````
Why use a query parameter instead of just using `/templates/<name>`?

jsjoeio replied:
I don't think we had originally considered it. The current approach doesn't care about the `/templates` path from the API the user interacts with. One downside would be that we're tying it to the templates path and if we change it, we'd have to support the old approach and the new approach.

Probably not a huge deal and your suggestion makes more sense considering our original [assumptions](https://github.com/coder/coder/issues/3981#issuecomment-1247186536). 

Should I change it?

kylecarbs replied:
I think changing it would make sense!

bpmct replied:
Agree with changing it to `/templates/<name>`

bpmct replied:
I'm a bit confused about what this PR does given links to `/templates/<name>` can be used to go to the template page.

jsjoeio replied:
Cool! I agree, makes sense to change.

> I'm a bit confused about what this PR does given links to /templates/<name> can be used to go to the template page.

Guess I'm confused too 😂 @bpmct I thought we agreed to <deployment-url> + <query params> in [our original spec](https://github.com/coder/coder/issues/3981#issuecomment-1247186536)?

bpmct replied:
Yeah - a good learning experience for me, my bad.

In v1, we have a separate path/flow for "Open in Coder" so it's not necessarily a problem that it is on a separate path, but I don't think we should stick to an original spec when a better route cames up, even if it results in some churn. The reason it was on a separate path for v1 was to handle Git OAuth a / different flow which wasn't addressed here. If we anticipate we'll do that down the road, I can see the merit of doing it that way.

Secondly, in its current state, the feature felt more like debt versus a UX improvement since the same value can be provided as a link to the templates page. If this provided additional functionality and a second route was necessary, I think it would have made more sense to keep.

jsjoeio replied:
Ah, that make sense. 

No worries at all! I agree, using the templates path makes more sense here and won't introduce any debt. I don't want swaying from the spec. Just wanted to make sure I didn't miss anything and also making sure we're on the same page 👍🏼 I'll make these changes then re-request a review.

presleyp reviewed (approved):
The machine looks good! I noticed that if I fake an SSE error, I see "an unknown error has occurred". This can be for a later PR but it would be good to have an informative error message.

jsjoeio commented:
> I noticed that if I fake an SSE error, I see "an unknown error has occurred". This can be for a later PR but it would be good to have an informative error message.

I'm less familiar with how that should work so I'd love to pair on it in a later PR if you're up for it!

jsjoeio committed:
Revert "feat: support template query param in index"

This reverts commit bad7ffb67718faea82aca7431ddf671c6a0e9410.

We decided to use the `/template/path` route instead.

jsjoeio committed:
fixup!: docs: add open in coder

jsjoeio committed:
docs: add open in coder to dogfood readme

jsjoeio commented:
@bpmct ready for another review!

bpmct reviewed (commented):
What's the plan for the additional functionality (e.g. autofill query parameters, detect if workspace exists, etc)? Are you planning on doing that after this in a separate PR?

jsjoeio committed:
Update docs/admin/open-in-coder.md

Co-authored-by: Ben Potter <ben@coder.com>

jsjoeio commented on docs/admin/open-in-coder.md:
````diff
@@ -0,0 +1,29 @@
+
+A basic example looks like this:
+
+```text
````
I thought if I used `markdown` as the lang, the website would render as Markdown instead of showing the code but I don't think that's the case so I'll change this.

jsjoeio committed:
Update docs/admin/open-in-coder.md

Co-authored-by: Ben Potter <ben@coder.com>

jsjoeio commented on docs/admin/open-in-coder.md:
```diff
@@ -0,0 +1,29 @@
+# Open in Coder Button
+
+> WARNING: this feature in Alpha and the API is subject to change.
```
```suggestion
```

jsjoeio committed:
Update docs/admin/open-in-coder.md

jsjoeio commented:
> What's the plan for the additional functionality (e.g. autofill query parameters, detect if workspace exists, etc)? Are you planning on doing that after this in a separate PR?

Yes, exactly! Next PR after this will be `createWorkspace`, implemented per the spec.
